### PR TITLE
feat(metal/bf16): additional cast support for bf16

### PR DESCRIPTION
**Pull request by kyle-mccarthy**
- extended bf16 cast support and testing
- added tests for bf16 binary ops
- added `is_u32_bf16` and `is_u8_bf16`
- exposed new + some existing kernels to candle-core's metal backend

### CODE REVIEW

File: candle-core/Cargo.toml [hunk 1]
```diff
@@ -48,4 +48,4 @@ metal = ["dep:metal", "dep:candle-metal-kernels"]
 [[bench]]
 name = "matmul"
 harness = false
-
+required-features = ["metal"]
```
Comment (LaurentMazare) [inline candle-core/Cargo.toml:51]: I don't think we want metal by default here.

Comment (kyle-mccarthy) [reply candle-core/Cargo.toml:51]: The [benchmark calls new_metal](https://github.com/huggingface/candle/blob/fa3ea98ba92835960fdd825a5b4dda30ef2baaa4/candle-core/benches/matmul.rs#L15) which is [only available when the metal feature is enabled](https://github.com/huggingface/candle/blob/fa3ea98ba92835960fdd825a5b4dda30ef2baaa4/candle-core/src/lib.rs#L92-L94).

Comment (ivarflakstad) [reply candle-core/Cargo.toml:51]: Yes, and we'll have to fix that.
I have a branch locally where I conditionally load the correct benchmarks from different modules via `cfg`.
I can make a PR tomorrow.

Comment (kyle-mccarthy) [reply candle-core/Cargo.toml:51]: Okay I figured that was the original intention. FWIW, rust-analyzer treats it as an error currently.

> error[E0599]: no method named `wait_until_completed` found for reference `&MetalDevice` in the current scope
  --> candle-core/benches/matmul.rs:31:24
   |
31 |                 device.wait_until_completed().unwrap();
   |                        ^^^^^^^^^^^^^^^^^^^^ method not found in `&MetalDevice`

Comment (ivarflakstad) [reply candle-core/Cargo.toml:51]: Something like this?
https://github.com/huggingface/candle/pull/1538

Comment (Narsil) [reply candle-core/Cargo.toml:51]: I kind of agree with should keep this out. Also the bench will be useful for other targets other than metal.

Comment (kyle-mccarthy) [reply candle-core/Cargo.toml]: okay it has been removed from the PR!

**Comment by kyle-mccarthy** [comment]
Is there any interest in merging PR? If so, I can fix the existing merge conflicts.

**Comment by Narsil** [comment]
ofc it's interesting.

**Pull request merged**